src/components/rules-creation-page/steps/IncludeExcludeSelectionStep.tsx

```diff
@@ -0,0 +1,234 @@
+    "organisationUnitGroups.organisationUnitGroupSets.attributes":
+        "Attributes of organisation unit group sets in organisation unit groups",
+    organisationUnitGroupSets: "Organisation unit group sets",
+    "organisationUnitGroupSets.attributes": "Attributes of organisation unit group sets",
+    users: "Users",
+    validationRules: "Validation rules",
+    "validationRules.attributes": "Attributes of validation rules",
+    "validationRules.validationRuleGroups": "Validation rule groups of validation rules",
+    validationRuleGroups: "Validation rule groups",
+    "validationRuleGroups.attributes": "Attributes of validation rule groups",
+};
+
+const IncludeExcludeSelectionStep: React.FC<IncludeExcludeSelectionStepProps> = ({
+    d2,
+    syncRule,
+    onChange,
+}) => {
+    const [modelSelectItems, setModelSelectItems] = useState<ModelSelectItem[]>([]);
+    const [models, setModels] = useState<typeof D2Model[]>([]);
+    const [selectedType, setSelectedType] = useState<string>("");
+
+    useEffect(() => {
+        getMetadata(getBaseUrl(d2), syncRule.metadataIds, "id,name").then((metadata: any) => {
+            const models = _.keys(metadata).map((type: string) => {
+                return d2ModelFactory(d2, type);
+            });
+
+            const parseModels = (models: typeof D2Model[]) =>
+                models
+                    .map((model: typeof D2Model) => model.getD2Model(d2))
+                    .map((model: ModelDefinition) => ({
+                        name: model.displayName,
+                        id: model.name,
+                    }));
+
+            setModels(models);
+            setModelSelectItems(parseModels(models));
+        });
+
+        console.log(syncRule.metadataExcludeIncludeRules);
+    }, [d2, syncRule]);
+
+    const changeUseDefaultIncludeExclude = (useDefault: boolean) => {
+        onChange(
+            useDefault
+                ? syncRule.markToUseDefaultIncludeExclude()
+                : syncRule.markToNotUseDefaultIncludeExclude(models)
+        );
+    };
+
+    const changeModelName = (event: any) => {
+        setSelectedType(event.target.value);
+    };
+
+    const changeInclude = (includeRules: any) => {
+        const type: string = selectedType;
+
+        const oldIncludeRules = getIncludeRules();
+        const oldExcludeRules = getExcludeRules();
+
+        const ruleIndexesToExclude = _.difference(oldIncludeRules, includeRules).map(rule =>
+            oldIncludeRules.indexOf(rule)
+        );
+        const ruleIndexesToInclude = _.difference(includeRules, oldIncludeRules).map(rule =>
+            oldExcludeRules.indexOf(rule)
+        );
+
+        if (ruleIndexesToInclude.length > 0) {
+            onChange(syncRule.moveRuleFromExcludeToInclude(type, ruleIndexesToInclude));
+        } else if (ruleIndexesToExclude.length > 0) {
+            onChange(syncRule.moveRuleFromIncludeToExclude(type, ruleIndexesToExclude));
```

Comment:
[food for thought] I'm not sure sending out the indexes to the model is the best approach here. I would've just sent the difference'd arrays instead and in the method use the rule instead of its index.

This could be a big change, so it's okay to me leave it as-is for now. But having to map-to-index to later undo the map with an includes(index) seems an overkill. Since we're sending the reference in function calls, the cost of sending the whole array or the indexes array is the same.

Reply:
Initially, I sent the strings of the rules but, what happens if they the sync rule receive a string of a rule that does not exist? throw an error? should the model allow this possible error? for this reason, use the indexes I think it is less prone to bugs

Reply:
Well actually with indexes we would face a worse behavior error. If it does not exist we would be selecting a different rule altogether and we wouldn't notice because there's no error handling.

Since we are building the list of possible rules, we should assume as a pre-condition that we send back correct strings that exist. Also using indexes makes the code/logic/debugging more cryptic.

Reply:
It is understood!! This could be a big change because it is necessary to refactor unit tests.
So, do I make this change now or when we move this feature to development-malaria?